### PR TITLE
Markdown comments

### LINKED ISSUE #2824: Rich text (segment) annotations
Tune webKnossos comment functionality to allow for rich text segment/node annotations:

  - [x] Virtualize comment list
  - [x] Render markdown for comments
  - [x] Allow to link to nodes/positions in comments, maybe using `#<nodeid>` and `#<x,y,z>`

### PR DESCRIPTION
**Pull request by daniel-wer**
This PR virtualizes the comment list in the tracing view, enables markdown rendering for comments and adds clickable references to other nodes and positions in comments.
Multiline comments are shown in proper markdown in a small popover next to the comment, when active.
I've also added a modal to enable proper markdown editing in a multiline textarea (click on edit markdown button). The sorting in the comment tab was reworked as well, natural sort is no longer the default (too slow) and it is now possible to reverse the sorting order in all sort modes (not only when sorting by name as previously).
I'm not entirely happy with the UI. The display of very large comments in the popover looks out of place and the single line comment input is also not perfect. One potential problem I see there is that users hit "Enter" to create a new comment, which now creates a new line and makes users believe their comment was not saved (as they only see the new empty line). Any recommendations appreciated :)

@fm3 As mentioned today, the backend will need to rewrite references in comments, that reference other nodes (`#<nodeid>`), in case nodeIds are changed during a merge. The frontend code to do this does a simple regex replace on comments using the regex `/#([0-9]+)/g`. If the referenced id is invalid, not existing or unchanged, nothing should be replaced.
I've also noticed that line breaks are stripped from comments when uploading an NML through the backend, could you check how to not do this :)

### URL of deployed dev instance (used for testing):
- https://markdowncomments.webknossos.xyz

### Steps to test:
- Upload NML with multiline comments and markdown using frontend/backend - The comments should be displayed correctly
- Create multiline comments with markdown using the edit markdown modal
- Reference other nodeIds (`#<nodeid>`) or positions (`#(<x,y,z>)`) which should be clickable and activate the referenced node / set the referenced position on click
- Also test Navigation to Dashboard/Other pages and test task hot-swap (because I removed code in the url_manager that was originally needed for this functionality to work correctly - however, during my testing I didn't notice any problems)

### TODO
 - [ ]  Compatibility btw front-end and back-end XML parser/writer concerning line-breaks

### Issues:
- fixes #2824

------
- [ ] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- [x] Ready for review

### CODE REVIEW

**Comment by fm3** [comment]
I added the comment transformation when merging tracings in the backend.
Two general thoughts: 
- should we maybe match only whole words?
- should we exclude non-existing nodes from the transformation? Or are you already doing that?

**Comment by daniel-wer** [comment]
> should we exclude non-existing nodes from the transformation? Or are you already doing that?

If the referenced id is not existing or unchanged, the frontend doesn't replace it. If a referenced node id doesn't exist, it won't be changed during a merge anyways, so it wouldn't be possible to replace it (as there is now new id), or am I misunderstanding you?

> should we maybe match only whole words?

I thought about this as well, but came to the conclusion that it is rather tricky. `#123`, `Node #123, for example, ...`, `See node #123: There is ...` are all valid references, and there's more of course. Do you know a simple way to do this? :)

File: app/assets/javascripts/navbar.js [hunk 1]
```diff
@@ -178,7 +178,7 @@ class Navbar extends React.PureComponent<Props, State> {
       <Header style={navbarStyle}>
         <Menu
           mode="horizontal"
-          defaultSelectedkeys={[this.props.history.location.pathname]}
+          defaultSelectedKeys={[this.props.history.location.pathname]}
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply app/assets/javascripts/navbar.js]: Yes :tada: And it fixes the highlighting of the current page in the navbar :)

**Comment by fm3** [comment]
> If the referenced id is not existing or unchanged, the frontend doesn't replace it. If a referenced node id doesn't exist, it won't be changed during a merge anyways, so it wouldn't be possible to replace it (as there is now new id), or am I misunderstanding you?

You are right, I’ll double check that the backend handles this correctly.

> I thought about this as well, but came to the conclusion that it is rather tricky.

I know no easy way, so I guess it’s fair to not do that. The UI gives feedback on what is a link anyway, so that should be fine :)

**Comment by daniel-wer** [comment]
> You are right, I’ll double check that the backend handles this correctly.
Sweet :)

Could you also have a look at the stripped line breaks when uploading an NML with multiline comments through the backend?

File: app/assets/javascripts/oxalis/view/right-menu/comment_tab/comment_tab_view.js [hunk 2]
```diff
@@ -110,87 +177,205 @@ class CommentTabView extends React.PureComponent<Props, CommentTabStateType> {
...
-        <TreeCommentList
-          key={tree.treeId}
+  setMarkdownModalVisibility = (visible: boolean) => {
+    this.setState({ isMarkdownModalVisible: visible });
+  };
+
+  renderMarkdownModal() {
+    const activeCommentMaybe = this.getActiveComment();
+    const activeNodeMaybe = getActiveNode(this.props.skeletonTracing);
+    const commentMaybe = activeNodeMaybe.map(({ id }) => ({
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply app/assets/javascripts/oxalis/view/right-menu/comment_tab/comment_tab_view.js]: `activeCommentMaybe` is a nothing unless the active node has a comment. `commentMaybe` now is only nothing if there is no active node. It represents a possible comment (with possibly empty content). If the active node indeed has a comment, the content of the `commentMaybe` comment is set accordingly.
I like your suggestion of handling that inside the `getActiveComment` function via an optional parameter and will change it that way :)

**Comment by fm3** [comment]
I encountered a minor layouting problem (Firefox 61)

image

**Comment by fm3** [comment]
Lab feedback was that their XML implementations are okay with multi-line attributes. I will look into how that can be done in our backend as well. Not sure if we should block this PR until that is complete

File: app/assets/javascripts/oxalis/model/reducers/skeletontracing_reducer.js [hunk 3]
```diff
@@ -77,7 +77,7 @@ function SkeletonTracingReducer(state: OxalisState, action: ActionType): OxalisS
           const lastTree = Maybe.fromNullable(_.maxBy(_.values(trees), tree => tree.treeId));
           return lastTree.map(t => {
             // use last node for active node
-            const lastNode = _.maxBy(_.values(t.nodes), node => node.id);
+            const lastNode = _.maxBy(Array.from(t.nodes.values()), node => node.id);
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/model/reducers/skeletontracing_reducer.js]: :+1:

File: package.json [hunk 4]
```diff
@@ -155,6 +155,9 @@
     "tween.js": "^16.3.1",
     "whatwg-fetch": "^1.1.0"
   },
+  "resolutions": {
+    "**/mini-store": "^1.1.0"
```
Comment (philippotto) [inline package.json]: Awesome find :+1:

**Pull request merged**